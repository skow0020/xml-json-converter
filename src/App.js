import './App.css';

import React, { Component } from 'react';

import Button from '@material-ui/core/Button';
import Input from '@material-ui/core/Input';
import logo from './logo.svg';
import xmlParser from 'fast-xml-parser';

const Parser = require('fast-xml-parser').j2xParser;

class App extends Component {
  generateJSON() {
    const xmlElementContent = document.getElementById('xml-file-content').textContent;
    const isValid = xmlParser.validate(xmlElementContent);

    if (xmlElementContent === '') {
      alert('Select an xml document');
    } else if (isValid !== true) {
      alert(`Invalid XML: ${isValid.err.msg}`);
    } else {
      const element = document.getElementById('json-file-content');
      const tObj = xmlParser.getTraversalObj(xmlElementContent);
      const jsonObjConverted = xmlParser.convertToJson(tObj);
      element.hidden = false;
      element.textContent = JSON.stringify(jsonObjConverted, null, 2);
    }
  }

  generateXML() {
    const jsonElementContent = document.getElementById('json-file-content').textContent;
    const element = document.getElementById('xml-file-content');

    if (jsonElementContent === '') {
      alert('Select a json document');
      return;
    }

    const jsonParser = new Parser({
      format: true,
      indentBy: '  '
    });

    try {
      element.textContent = jsonParser.parse(JSON.parse(jsonElementContent));
      element.hidden = false;
    } catch (error) {
      alert(`Unable to generate json: ${error.message}`);
    }
  }

  readFile(e) {
    const file = e.target.files[0];
    let element;
    if (!file) return;
    if (file.type === 'application/json') element = document.getElementById('json-file-content');
    else if (file.type === 'text/xml') element = document.getElementById('xml-file-content');
    else {
      alert('File must be xml or json');
      return;
    }

    const reader = new FileReader();
    reader.onload = (g) => {
      const contents = g.target.result;
      element.hidden = false;
      element.textContent = contents;
    };
    reader.readAsText(file);
  }

  clearData() {
    document.getElementById('json-file-content').hidden = true;
    document.getElementById('json-file-content').textContent = '';
    document.getElementById('xml-file-content').hidden = true;
    document.getElementById('xml-file-content').textContent = '';
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
          <Input id="file-input" type="file" onChange={e => this.readFile(e)} />
          <div id="container">
            <Button id="generate-json" variant="contained" color="primary" onClick={this.generateJSON}>
              Generate JSON
            </Button>
            <Button id="generate-xml" variant="contained" color="primary" onClick={this.generateXML}>
              Generate XML
            </Button>
          </div>
          <Button id="clear-button" variant="contained" color="secondary" onClick={this.clearData}>
            Clearasdf
          </Button>
        </header>
        <div className="content-container">
          <pre id="xml-file-content" className="file-content" hidden />
          <pre id="json-file-content" className="file-content" hidden />
        </div>
      </div>
    );
  }
}

export default App;
