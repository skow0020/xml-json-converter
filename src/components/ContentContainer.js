import React, { Component } from 'react';

import PropTypes from 'prop-types';

class ContentContainer extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    className: PropTypes.string.isRequired,
    hidden: PropTypes.bool.isRequired,
    content: PropTypes.string.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
    };
  }

  render() {
    const { id, className, hidden, content } = this.props;
    return (
      <pre
        id={id}
        className={className}
        hidden={hidden}
      >
        {content}
      </pre>
    );
  }
}

export default ContentContainer;
