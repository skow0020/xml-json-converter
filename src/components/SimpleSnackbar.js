import CloseIcon from '@material-ui/icons/Close';
import IconButton from '@material-ui/core/IconButton';
import React from 'react';
import Snackbar from '@material-ui/core/Snackbar';
import { withStyles } from '@material-ui/core/styles';

const styles = theme => ({
  close: {
    padding: theme.spacing.unit / 2
  }
});

let openSnackbarFn;

class SimpleSnackbar extends React.Component {
  state = {
    open: false,
    message: ''
  };

  componentDidMount() {
    openSnackbarFn = this.openSnackbar;
  }

  openSnackbar = ({ message }) => {
    this.setState({
      open: true,
      message
    });
  };

  handleClose = (event, reason) => {
    if (reason === 'clickaway') {
      return;
    }

    this.setState({
      open: false,
      message: ''
    });
  };

  render() {
    const message = (
      <span
        id="snackbar-message-id"
        dangerouslySetInnerHTML={{ __html: this.state.message }}
      />
    );
    return (
      <div>
        <Snackbar
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'left'
          }}
          open={this.state.open}
          autoHideDuration={2000}
          onClose={this.handleClose}
          ContentProps={{
            'aria-describedby': 'message-id'
          }}
          message={message}
          action={[
            <IconButton
              key="close"
              aria-label="Close"
              color="inherit"
              onClick={this.handleClose}
            >
              <CloseIcon />
            </IconButton>
          ]}
        />
      </div>
    );
  }
}

export function openSnackbar({ message }) {
  openSnackbarFn({ message });
}

export default withStyles(styles)(SimpleSnackbar);
