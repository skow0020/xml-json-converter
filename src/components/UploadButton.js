import React, { Component } from 'react';

import Button from '@material-ui/core/Button';
import PropTypes from 'prop-types';

class UploadButton extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    className: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
    };
  }

  render() {
    const { id, className, onChange } = this.props;
    return (
      <label id="file-select" htmlFor={id}>
        <input
          style={{ display: 'none' }}
          id={id}
          type="file"
          name={className}
          onChange={onChange}
        />
        <Button variant="contained" component="span">Upload File</Button>
      </label>
    );
  }
}

export default UploadButton;
