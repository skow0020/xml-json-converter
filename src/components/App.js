import '../App.css';

import React, { Component } from 'react';
import SimpleSnackbar, { openSnackbar } from './SimpleSnackbar';

import Button from '@material-ui/core/Button';
import ContentContainer from './ContentContainer';
import UploadButton from './UploadButton';
import logo from '../logo.svg';
import xmlParser from 'fast-xml-parser';

const Parser = require('fast-xml-parser').j2xParser;

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      xmlContentHidden: true,
      jsonContentHidden: true,
      xmlContent: '',
      jsonContent: ''
    };

    this.clearData = this.clearData.bind(this);
    this.generateJSON = this.generateJSON.bind(this);
    this.generateXML = this.generateXML.bind(this);
    this.readFile = this.readFile.bind(this);
  }

  showNotifier = (message) => {
    openSnackbar({ message });
  }

  generateJSON() {
    const xmlElementContent = this.state.xmlContent;
    const isValid = xmlParser.validate(xmlElementContent);

    if (xmlElementContent === '') {
      this.showNotifier('Select an XML document');
    } else if (isValid !== true) {
      this.showNotifier(`Invalid XML: ${isValid.err.msg}`);
    } else {
      const tObj = xmlParser.getTraversalObj(xmlElementContent);
      const jsonObjConverted = xmlParser.convertToJson(tObj);

      this.setState(() => ({
        jsonContentHidden: false,
        jsonContent: JSON.stringify(jsonObjConverted, null, 2)
      }));
    }
  }

  generateXML() {
    const jsonElementContent = this.state.jsonContent;

    if (jsonElementContent === '') {
      this.showNotifier('Select a JSON Document');
      return;
    }

    const jsonParser = new Parser({
      format: true,
      indentBy: '  '
    });

    try {
      this.setState(() => ({
        xmlContentHidden: false,
        xmlContent: jsonParser.parse(JSON.parse(jsonElementContent))
      }));
    } catch (error) {
      this.showNotifier(`Unable to generate json: ${error.message}`);
    }
  }

  readFile(e) {
    const file = e.target.files[0];
    if (!file) return;
    if (file.type === 'application/json') {
      const reader = new FileReader();
      reader.onload = (g) => {
        this.setState(() => ({
          jsonContentHidden: false,
          jsonContent: g.target.result
        }));
      };
      reader.readAsText(file);
    } else if (file.type === 'text/xml') {
      const reader = new FileReader();
      reader.onload = (g) => {
        this.setState(() => ({
          xmlContentHidden: false,
          xmlContent: g.target.result
        }));
      };
      reader.readAsText(file);
    } else {
      this.showNotifier('File must be xml or json');
    }
  }

  clearData() {
    this.setState(() => ({
      xmlContentHidden: true,
      jsonContentHidden: true,
      xmlContent: '',
      jsonContent: ''
    }));
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
          <UploadButton id="contained-button-file" className="upload-button" onChange={e => this.readFile(e)} />
          <div id="generation-container">
            <Button id="generate-json" variant="contained" color="primary" onClick={() => this.generateJSON()}>Generate JSON</Button>
            <Button id="generate-xml" variant="contained" color="primary" onClick={() => this.generateXML()}>Generate XML</Button>
          </div>
          <Button id="clear-button" variant="contained" color="secondary" onClick={() => this.clearData()}>Clear Data</Button>
        </header>
        <SimpleSnackbar />
        <ContentContainer
          id="xml-file-content"
          className="file-content"
          hidden={this.state.xmlContentHidden}
          content={this.state.xmlContent}
        />
        <ContentContainer
          id="json-file-content"
          className="file-content"
          hidden={this.state.jsonContentHidden}
          content={this.state.jsonContent}
        />
      </div>
    );
  }
}

export default App;
