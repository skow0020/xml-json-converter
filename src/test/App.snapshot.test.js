import App from '../components/App';
import ContentContainer from '../components/ContentContainer';
import React from 'react';
import SimpleSnackbar from '../components/SimpleSnackbar';
import UploadButton from '../components/UploadButton';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('App Snapshot Tests', () => {
  it('App snapshot test', async () => {
    const wrapper = shallow(
      <App />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('ContentContainer hidden snapshot test', async () => {
    const wrapper = shallow(
      <ContentContainer
        id="xml-file-content"
        className="file-content"
        hidden
        content="<XML></XML>"
      />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('ContentContainer hidden=false snapshot test', async () => {
    const wrapper = shallow(
      <ContentContainer
        id="xml-file-content"
        className="file-content"
        hidden={false}
        content="<XML></XML>"
      />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('UploadButton snapshot test', async () => {
    const wrapper = shallow(
      <UploadButton id="contained-button-file" className="upload-button" onChange={() => { }} />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('SimpleSnackbar snapshot test', () => {
    const wrapper = shallow(<SimpleSnackbar open message="THIS IS A MESSAGE" />);
    wrapper.setState({
      open: true
    });

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
