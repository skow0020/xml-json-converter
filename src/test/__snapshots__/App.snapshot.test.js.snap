// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Snapshot Tests App snapshot test 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <label
      htmlFor="contained-button-file"
      id="file-select"
    >
      <input
        id="contained-button-file"
        onChange={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        type="file"
      />
      <span
        className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-contained-12 MuiButton-raised-15"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiButton-label-2"
        >
          Upload File
        </span>
        <span
          className="MuiTouchRipple-root-30"
        />
      </span>
    </label>
    <div
      id="container"
    >
      <button
        className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16"
        disabled={false}
        id="generate-json"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-2"
        >
          Generate JSON
        </span>
        <span
          className="MuiTouchRipple-root-30"
        />
      </button>
      <button
        className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16"
        disabled={false}
        id="generate-xml"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-2"
        >
          Generate XML
        </span>
        <span
          className="MuiTouchRipple-root-30"
        />
      </button>
    </div>
    <button
      className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-contained-12 MuiButton-containedSecondary-14 MuiButton-raised-15 MuiButton-raisedSecondary-17"
      disabled={false}
      id="clear-button"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-2"
      >
        Clear Data
      </span>
      <span
        className="MuiTouchRipple-root-30"
      />
    </button>
  </header>
  <div
    className="content-container"
  >
    <pre
      className="file-content"
      hidden={true}
      id="xml-file-content"
    >
      
    </pre>
    <pre
      className="file-content"
      hidden={true}
      id="json-file-content"
    >
      
    </pre>
  </div>
</div>
`;
